### PR TITLE
Remove oc distributed products n+1

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

Related to https://github.com/openfoodfoundation/openfoodnetwork/issues/3853

#### Release notes

Changelog Category: Added | Changed | Deprecated | Removed | Fixed | Security

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
The de-N-plus-one-pocalypse has begun! :tada:

**Comment by sauloperez** [comment]
LOL reminds me 



and I'm just following your command. You created the issue :joy:

File: app/services/order_cycle_distributed_products.rb [hunk 1]
```diff
@@ -15,14 +15,25 @@ def initialize(order_cycle, distributor)
...
+      .merge(order_cycle.variants_distributed_by(distributor))
+
+    Rails.logger.debug "====== variants ======="
     variants = order_cycle.variants_distributed_by(distributor)
-    products = variants.map(&:product).uniq
 
+    Rails.logger.debug "====== reject ======="
     valid_products = products.reject do |product|
       product_has_only_obsolete_master_in_distribution?(product, variants)
     end
```
Comment (sauloperez) [inline app/services/order_cycle_distributed_products.rb]: It is this block that causes the N+1s

Comment (Matt-Yorkley) [reply app/services/order_cycle_distributed_products.rb]:
```ruby
hundreds_of.products do |n|
   n+1
end
```
Classic.

Comment (Matt-Yorkley) [reply app/services/order_cycle_distributed_products.rb]: Yeah, I remember digging around in this code for a while trying some basic eager-loading solutions, but didn't manage to crack it.

This whole section looks pretty mad:
https://github.com/openfoodfoundation/openfoodnetwork/blob/59d4e3d9e2192524d807b8dbc4271b427adea1ac/app/services/order_cycle_distributed_products.rb#L17-L27

There's multiple queries of products and variants here in short succession. I was wondering if we could make it into a single complex query, with some fancy `#merge` magic?

Comment (Matt-Yorkley) [reply app/services/order_cycle_distributed_products.rb]: Am I right in thinking there's another query to variants inside of `#product_has_only_obsolete_master_in_distribution`? So we get variants, get products of those variants, get fetch the variants of those products, and then finally query products again? Something like that?

Comment (sauloperez) [reply app/services/order_cycle_distributed_products.rb]: Yep, exactly like that @Matt-Yorkley  

> I was wondering if we could make it into a single complex query, with some fancy #merge magic?

That's what I'm trying but it's painfully hard for me to turn imperative code to declarative :expressionless: :boom:

**Pull request closed without merging**